### PR TITLE
Add WeightToFee and LengthToFee impls to transaction-payment Runtime API

### PR DESCRIPTION
**Pull request by notlesh**
This PR makes a runtime's `WeightToFee` and `LengthToFee` impls available via runtime API. The rationale for this is that these are now opaque functions which are not directly available in any way outside of a runtime.

Specifically, a client cannot take an executed extrinsic from a block and validate each portion of its fees without knowing this information (or making some assumptions). We do this extensively in testing on Moonbeam.

polkadot companion: https://github.com/paritytech/polkadot/pull/6536
cumulus companion: https://github.com/paritytech/cumulus/pull/2073

### CODE REVIEW

File: frame/transaction-payment/rpc/src/lib.rs [hunk 1]
```diff
@@ -48,6 +48,20 @@ pub trait TransactionPaymentApi<BlockHash, ResponseType> {
...
+		weight: sp_weights::Weight,
+		at: Option<BlockHash>,
+	) -> RpcResult<NumberOrHex>;
+
+	#[method(name = "payment_lengthToFee")]
+	fn query_length_to_fee(
+		&self,
+		length: u32,
+		at: Option<BlockHash>,
+	) -> RpcResult<NumberOrHex>;
```
Comment (bkchr) [inline frame/transaction-payment/rpc/src/lib.rs:64]: New RPCs are discouraged. Having the runtime api is enough. People can call the state_call rpc to call the api.

**Review by bkchr — COMMENTED** [review]
I think in general it is okay. However, we should not add a new RPC.

File: frame/transaction-payment/rpc/src/lib.rs (hunk 1 above)
Comment (notlesh) [reply frame/transaction-payment/rpc/src/lib.rs]: Good point, RPC seems unnecessary. Removed in ee6f399.

File: frame/transaction-payment/rpc/runtime-api/src/lib.rs [hunk 2]
```diff
@@ -46,5 +48,11 @@ sp_api::decl_runtime_apis! {
 
 		/// Query fee details of a given encoded `Call`.
 		fn query_call_fee_details(call: Call, len: u32) -> FeeDetails<Balance>;
+
+		/// Query the output of the current WeightToFee given some input
+		fn query_weight_to_fee(weight: sp_weights::Weight) -> Balance;
+
+		/// Query the output of the current LengthToFee given some input
+		fn query_length_to_fee(length: u32) -> Balance;
```
Comment (joepetrowski) [inline frame/transaction-payment/rpc/runtime-api/src/lib.rs:56]:
```suggestion
		/// Query the output of the current `WeightToFee` given some input.
		fn query_weight_to_fee(weight: sp_weights::Weight) -> Balance;

		/// Query the output of the current `LengthToFee` given some input.
		fn query_length_to_fee(length: u32) -> Balance;
```

File: frame/transaction-payment/src/lib.rs [hunk 3]
```diff
@@ -607,11 +607,11 @@ where
 		}
 	}
 
-	fn length_to_fee(length: u32) -> BalanceOf<T> {
+	pub fn length_to_fee(length: u32) -> BalanceOf<T> {
 		T::LengthToFee::weight_to_fee(&Weight::from_ref_time(length as u64))
 	}
 
-	fn weight_to_fee(weight: Weight) -> BalanceOf<T> {
+	pub fn weight_to_fee(weight: Weight) -> BalanceOf<T> {
```
Comment (bkchr) [inline frame/transaction-payment/src/lib.rs:614]: Can you please add some docs.

**Review by bkchr — APPROVED** [review]

**Review by joepetrowski — APPROVED** [review]

**Comment by notlesh** [comment]
In trying to test this, I came across something unexpected. With my changes (basically [my Polkadot companion PR](https://github.com/paritytech/polkadot/pull/6536) plus some diener-created overrides to this PR), there is no `transactionPaymentApi` or `transactionPaymentCallApi` listed under `developer -> Runtime calls`:

image

When compiling without my changes (`polkadot` `master`), they appear as expected:

image

It seems others are missing as well, such as `accountNonceApi`.

Any idea what might cause this?

**Comment by notlesh** [comment]
In trying to test this, I came across something unexpected. With my changes ([my Polkadot companion PR](https://github.com/paritytech/polkadot/pull/6536) plus diener overrides), the `transactionPaymentApi` and `transactionPaymentCallApi` both disappear from `Developer > Runtime calls -> Calls`:

image

On recent Polkadot `master`, they both appear as expected:

image

Any idea what might be causing this?

File: frame/transaction-payment/src/lib.rs (hunk 3 above)
Comment (notlesh) [reply frame/transaction-payment/src/lib.rs]: Added in 9767e6b

**Comment by notlesh** [comment]
> Any idea what might be causing this?

I believe this is because the type info is currently hard-coded here: https://github.com/polkadot-js/api/blob/ebb9f055c13d663bf3af5fad5043c5461b2f2b8a/packages/api-augment/src/substrate/runtime.ts#L268

Should I open a companion PR to include this new definition there? CC @jacogr

Are there examples of using either the `state_call` (I think that's the `State::call()` RPC) or `api.runtime` (the later doesn't appear in my testing with `@polkadot/api 9.11.2`).

**Comment by jacogr** [comment]
Indeed. At this point they need to be hard-coded on the JS API until such time as these appear in the metadata. Once there, it appears on `api.call.*` (assuming the runtime api version matches)

Alongside that, I would really encourage you to bump the actual runtime api version - because we are hardcoding these each version defines a specific shape (and the types). Unless we bump it all these will also appear decorated for runtimes that don't actually have them but has the same version. 

(Because of that compatibility issue, the JS API won't be adding them unless there is a version bump since it just ends up with masses of support otherwise when we have mismatched "shapes" of the same version)

**Comment by notlesh** [comment]
All companion PRs are now approved (the polkadot-js one approved verbally :) Let me know if there is anything missing :+1:

**Review by ggwpez — APPROVED** [review]

**Comment by ggwpez** [comment]
Dont know why the CI is read, lets try rebasing.

**Comment by ggwpez** [comment]
bot merge

**Comment by ggwpez** [comment]
bot merge

**Comment by ggwpez** [comment]
bot merge

**Comment by michalkucharczyk** [comment]
Companions were not merged.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v0-9-38/2122/1

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/substrate-developer-newsletter-february-2023-notable-updates-for-frame-devs-new-deep-dives-released-calling-builders-to-apply-to-rfps-and-bounties/2137/1

**Pull request merged**